src/System.Linq/src/System/Linq/Take.SpeedOpt.cs

```diff
@@ -12,5 +13,58 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(count > 0);
+
+            if (source is IPartition<TSource> partition)
+            {
+                int length = partition.GetCount(true);
+                if (length > 0)
+                {
+                    return length - count > 0 ? partition.Skip(length - count) : partition;
+                }
+                else if (length == 0)
+                {
+                    return partition;
+                }
+            }
+            else if (source is IList<TSource> sourceList)
+            {
+                if (sourceList.Count > count)
+                {
+                    return new ListPartition<TSource>(sourceList, sourceList.Count - count, sourceList.Count);
+                }
+                else
+                {
+                    return new ListPartition<TSource>(sourceList, 0, sourceList.Count);
+                }
+            }
+            else if (source is ICollection<TSource> collection)
+            {
+                if (collection.Count > count)
+                {
+                    return new EnumerablePartition<TSource>(collection.Skip(collection.Count - count), 0, count);
+                }
+                else
+                {
+                    return new EnumerablePartition<TSource>(collection, 0, collection.Count);
+                }
+            }
+            else if (source is IReadOnlyCollection<TSource> roCollection)
```

Comment:
We've explicitly not been adding in checks for IReadOnlyCollection.  We shouldn't start now in this PR.

Reply:
(To elaborate, we are avoiding using `IReadOnlyCollection<T>` until the runtime performance around typecasting to covariant interfaces improves significantly.  If you search in issues in coreclr/corefx, you should find a bunch of related issues.)

Reply:
Right. I noticed that, but assumed that was already reviewed and covariant interface casts are no longer a problem.

Reply:
> assumed that was already reviewed and covariant interface casts are no longer a problem

Alas, https://github.com/dotnet/coreclr/issues/603 remains open.

Reply:
Should this be somehow marked/uncommented or just removed?

Reply:
Just removed. Thanks.

Reply:
Done